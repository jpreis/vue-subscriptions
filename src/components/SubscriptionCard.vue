<template>
  <div class="subscription-card" :class="{inactive: !subscriptionActive}" @click="toggleActive" :style="backgroundStyle" @mouseenter="hovered = true" @mouseleave="hovered = false">

    <div class="name">
      {{ subscription.name }}
    </div>
    <div class="costs">
      {{ subscription.costs | dollars}}
    </div>

    <div class="controls" :class="{visible: hovered}">
      <ColorPicker :defaultColor="cardBackgroundColor" @color-selected="setCardBackgroundColor"></ColorPicker>
    </div>
  </div>
</template>

<script>
import ColorPicker from "@/components/ColorPicker.vue";
import { dollars } from "@/filter/CurrencyFilter";

export default {
  props: {
    subscription: {
      reqquired: true
    }
  },
  data: function() {
    return {
      cardBackgroundColor: "#445465",
      hovered: false
    };
  },
  filters: {
    dollars
  },
  components: {
    ColorPicker
  },
  methods: {
    setCardBackgroundColor(newColor) {
      this.cardBackgroundColor = newColor;
    },
    toggleActive() {
      this.$emit("toggle-active");
    }
  },
  computed: {
    subscriptionActive() {
      return this.subscription.active;
    },
    backgroundStyle() {
      return {
        "background-color": this.cardBackgroundColor
      };
    }
  }
};
</script>

<style lang="scss" scoped>
.subscription-card {
  transition: all 0.3s;
  padding: 1em 1.5em;
  color: white;
  box-shadow: 0 14px 28px rgba(0, 0, 0, 0.12), 0 10px 10px rgba(0, 0, 0, 0.06);
  position: relative;
  border-radius: 5px;

  &:hover {
    cursor: pointer;
  }

  &.inactive {
    opacity: 0.6;
    transform: translateY(5px);
    box-shadow: 0 7px 14px rgba(0, 0, 0, 0.075), 0 5px 5px rgba(0, 0, 0, 0.06);
    background: repeating-linear-gradient(
      -45deg,
      rgba(255, 255, 255, 0.1),
      rgba(255, 255, 255, 0.1) 10px,
      transparent 10px,
      transparent 20px
    );
  }
}

.name {
  text-transform: uppercase;
  font-size: 1.25rem;
  letter-spacing: 0.08em;
}

.costs {
  margin-top: 6rem;
  text-align: right;
  font-size: 2.5rem;
}

.controls {
  transition: opacity 0.3s;
  position: absolute;
  top: 1em;
  right: 1em;
  opacity: 0;
  display: grid;
  grid-auto-flow: row;
  grid-gap: 5px;
  align-items: center;

  &.visible {
    opacity: 1;
  }
}
</style>