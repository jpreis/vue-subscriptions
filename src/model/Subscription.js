export default class Subscription {
  static nextId = 0;

  constructor(name, costs, active = true) {
    this.id = Subscription.nextId++;
    this.name = name;
    this.costs = costs;
    this.active = active;
  }
}
