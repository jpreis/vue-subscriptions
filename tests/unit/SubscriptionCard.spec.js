import { shallowMount } from "@vue/test-utils";
import SubscriptionCard from "@/components/SubScriptionCard.vue";
import Subscription from "@/model/Subscription";

describe("SubscriptionCard.vue", () => {
  const subscription = new Subscription("Netflix", 9.99);
  let wrapper = null;

  beforeEach(() => {
    wrapper = shallowMount(SubscriptionCard, {
      propsData: { subscription }
    });
  });

  it("has the expected HTML structure", () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it("triggers a 'toggle-active' event when clicked", () => {
    wrapper.trigger("click");
    expect(wrapper.emitted("toggle-active")).toBeTruthy();
  });
});
