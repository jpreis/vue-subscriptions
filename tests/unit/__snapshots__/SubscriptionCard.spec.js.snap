// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubscriptionCard.vue has the expected HTML structure 1`] = `
<div
  class="subscription-card"
  style="background-color: rgb(68, 84, 101);"
>
  <div
    class="name"
  >
    
    Netflix
  
  </div>
   
  <div
    class="costs"
  >
    
    $9.99
  
  </div>
   
  <div
    class="controls"
  >
    <!---->
  </div>
</div>
`;
